### PR TITLE
Updated minSdkVersion to 21 and removed special cases for APIs under 21 

### LINKED ISSUE #3624: Set minSdkVersion to 21 and remove special cases for APIs under 21
The biggest change coming to the upcoming v1.26 release is starting the [scoped storage migration](https://forum.opendatakit.org/t/collect-will-need-to-stop-using-sdcard-odk-for-files/24159). This is not relevant to users on older versions of Android. Additionally, Android <5 now only accounts for 6% of Collect users.

The [Technical Steering Committee](https://opendatakit.org/community/governance/#tsc-1) discussed dropping support for Android < 5 and decided to move forward in this release.

There are currently several special cases for APIs under 21 and these should be removed.

### PR DESCRIPTION
**Pull request by SaumiaSinghal**
Closes #3624 

#### What has been done to verify that this works as intended?
I tested it.

#### Why is this the best possible solution? Were any other approaches considered?
I updated the minSdkVersion to 21, and clearing the codebase of special cases for target APIs under 21.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
No regression risks

#### Do we need any specific form for testing your changes? If so, please attach one.
No specific form

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/utilities/PRNGFixes.java [hunk 1]
```diff
@@ -39,107 +36,13 @@
  */
 public final class PRNGFixes {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/utilities/PRNGFixes.java]: Isn't this whole class unused now?

**Comment by grzesiek2010** [comment]
Please take a look at resources as well we have some folders for specific APIs like `layout-v17` for example.

File: collect_app/src/main/java/org/odk/collect/android/widgets/ImageWidget.java [hunk 2]
```diff
@@ -150,11 +148,7 @@ private void captureImage() {
         errorTextView.setVisibility(View.GONE);
         Intent intent;
         if (selfie) {
-            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
-                intent = new Intent(getContext(), CaptureSelfieActivityNewApi.class);
-            } else {
-                intent = new Intent(getContext(), CaptureSelfieActivity.class);
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/ImageWidget.java:162]: Looks like `CaptureSelfieActivity` should be removed at all because it's unused.

File: collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java [hunk 3]
```diff
@@ -352,26 +285,11 @@ public void denied() {
     private void captureVideo() {
         Intent i;
         if (selfie) {
-            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
-                i = new Intent(getContext(), CaptureSelfieVideoActivityNewApi.class);
-            } else {
-                i = new Intent(getContext(), CaptureSelfieVideoActivity.class);
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java:359]: The same here, looks as if this class was used just here so we can get rid of it.

**Comment by grzesiek2010** [comment]
Please search for annotations like `@TargetApi(Build.VERSION_CODES.LOLLIPOP)` since LOLLIPOP will be the min api we can get rid of it.

File: collect_app/src/main/java/org/odk/collect/android/utilities/PRNGFixes.java (hunk 1 above)
Comment (SaumiaSinghal) [reply collect_app/src/main/java/org/odk/collect/android/utilities/PRNGFixes.java]: Yes, actually @grzesiek2010! :sweat_smile:

File: collect_app/src/main/res/layout-v17/range_widget_horizontal.xml [hunk 4]
```diff
@@ -1,67 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (grzesiek2010) [inline collect_app/src/main/res/layout-v17/range_widget_horizontal.xml:1]: If you have resources with specified API level like here v17 it's not only for API 17 but for 17+. I know it's misleading but actually this file is the one we want to keep, it should be moved to general layout folder to replace the one that exists there and then this folder should be removed.

**Comment by grzesiek2010** [comment]
@SaumiaSinghal 
there are also resources with -21 suffix you should remove them, but make sure you understand how that works (look my comment above: https://github.com/opendatakit/collect/pull/3625#discussion_r376316352)

File: collect_app/src/main/res/layout-v17/range_widget_horizontal.xml (hunk 4 above)
Comment (SaumiaSinghal) [reply collect_app/src/main/res/layout-v17/range_widget_horizontal.xml:1]: Okay yes I understand why testDebug was not able to locate the file `range_widget_horizontal.xml`

**Comment by SaumiaSinghal** [comment]
Thanks for explaining so clearly @grzesiek2010. I think, all the cases for API under 21 have been removed now. Please let me know if something is still left.

File: collect_app/src/main/java/org/odk/collect/android/utilities/ViewIds.java [hunk 5]
```diff
@@ -1,47 +1,17 @@
 package org.odk.collect.android.utilities;
 
-import android.os.Build;
 import android.view.View;
 
-import java.util.concurrent.atomic.AtomicInteger;
-
 public class ViewIds {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/utilities/ViewIds.java]: Now it doesn't make sense to keep this class. Could you get rid of it and use `View.generateViewId()` everywhere instead?

File: collect_app/src/main/java/org/odk/collect/android/widgets/ImageWidget.java [hunk 6]
```diff
@@ -64,6 +62,7 @@ public ImageWidget(Context context, final QuestionDetails prompt) {
         addAnswerView(answerLayout);
     }
 
+
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/ImageWidget.java]: Please get rid of this extra empty line. Hmm it should break the build, i'm wondering why it didn't happen in this case.

File: collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieActivityNewApi.java [hunk 7]
```diff
@@ -28,7 +26,6 @@
 
 import static org.odk.collect.android.utilities.PermissionUtils.isCameraPermissionGranted;
 
-@TargetApi(Build.VERSION_CODES.LOLLIPOP)
 public class CaptureSelfieActivityNewApi extends CollectAbstractActivity {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieActivityNewApi.java]: Please rename this class removing `NewApi` because now it's misleading.

File: collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieVideoActivityNewApi.java [hunk 8]
```diff
@@ -28,7 +26,6 @@
 
 import static org.odk.collect.android.utilities.PermissionUtils.areCameraAndRecordAudioPermissionsGranted;
 
-@TargetApi(Build.VERSION_CODES.LOLLIPOP)
 public class CaptureSelfieVideoActivityNewApi extends Activity {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieVideoActivityNewApi.java]: Please rename this class removing `NewApi` because now it's misleading.

File: collect_app/src/main/java/org/odk/collect/android/widgets/RangeWidget.java [hunk 9]
```diff
@@ -172,14 +170,10 @@ private void setUpSeekBar() {
         seekBar.setOnSeekBarChangeListener(this);
         if (isRTL()) {
             float rotate = seekBar.getRotation();
-            if (Build.VERSION.SDK_INT == Build.VERSION_CODES.KITKAT ||
-                    Build.VERSION.SDK_INT == Build.VERSION_CODES.JELLY_BEAN_MR2) {
+            if (Build.VERSION.SDK_INT == Build.VERSION_CODES.KITKAT) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/RangeWidget.java]: There is something wrong here, `KITKAT` won't be supported so you don't need this check.

File: collect_app/src/main/res/values-v21/themes.xml [hunk 10]
```diff
@@ -1,19 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
-
-<!--
-    This file contains the app's light and dark theme definitions for
-    API 21 and above
```
Comment (grzesiek2010) [inline collect_app/src/main/res/values-v21/themes.xml:5]: Why did you remove this file? The comment is clear and I explained above:

> if you have resources with specified API level like here v17 it's not only for API 17 but for 17+. I know it's misleading but actually this file is the one we want to keep, it should be moved to general layout folder to replace the one that exists there and then this folder should be removed.

File: collect_app/src/main/res/drawable-v21/action_bar_shadow.xml [hunk 11]
```diff
@@ -1,20 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (grzesiek2010) [inline collect_app/src/main/res/drawable-v21/action_bar_shadow.xml:1]: Again we need this file it's for api 21+.

File: collect_app/src/main/res/layout-v21/toolbar_action_bar_shadow.xml [hunk 12]
```diff
@@ -1,6 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (grzesiek2010) [inline collect_app/src/main/res/layout-v21/toolbar_action_bar_shadow.xml:1]: We need this file it's for api 21+. You just need to move it to the main layout package to replace the one for other APIs.

**Review by seadowg — CHANGES_REQUESTED** [review]
Just sticking this into the "Changes requested" state based on @grzesiek2010's comments

File: collect_app/src/main/java/org/odk/collect/android/utilities/ViewIds.java (hunk 5 above)
Comment (SaumiaSinghal) [reply collect_app/src/main/java/org/odk/collect/android/utilities/ViewIds.java]: Yes @grzesiek2010, I will do that!

**Comment by SaumiaSinghal** [comment]
Hie @grzesiek2010. Thanks for reviewing! I was a little confused regarding the files needed for api 21+. I reverted a few changes. I hope it is fine now.

File: collect_app/src/main/java/org/odk/collect/android/fragments/AppListFragment.java [hunk 13]
```diff
@@ -104,12 +103,10 @@ public static void toggleButtonLabel(Button toggleButton, ListView lv) {
     public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
 
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
-            // Use the nicer-looking drawable with Material Design insets.
-            ListView listView = getListView();
-            listView.setDivider(getResources().getDrawable(R.drawable.list_item_divider, getActivity().getTheme()));
-            listView.setDividerHeight(1);
-        }
+        // Use the nicer-looking drawable with Material Design insets.
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/fragments/AppListFragment.java]: Probably don't need this comment any longer!

File: collect_app/src/main/java/org/odk/collect/android/utilities/MediaUtils.java [hunk 14]
```diff
@@ -419,32 +418,7 @@ public static final int deleteVideoInFolderFromMediaProvider(File folder) {
      */
     @SuppressLint("NewApi")
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/utilities/MediaUtils.java]: We can probably lose this `@SupressLint`

File: collect_app/src/main/res/layout-v17/range_widget_horizontal.xml (hunk 4 above)
Comment (seadowg) [reply collect_app/src/main/res/layout-v17/range_widget_horizontal.xml:1]: I think we're still missing this one unless I just can't see it?

**Review by seadowg — CHANGES_REQUESTED** [review]
Just a couple comments on the changes and it looks like there is a conflict - hard to avoid given how big the changes are annoyingly!

File: collect_app/src/main/res/values/themes.xml [hunk 15]
```diff
@@ -2,12 +2,18 @@
 
 <!--
     This file contains the app's light and dark theme definitions for
-    API 20 and below
+    API 21 and above
```
Comment (grzesiek2010) [inline collect_app/src/main/res/values/themes.xml]: Since now we won't support lower versions you can get rid of this comment.

File: collect_app/src/main/res/layout-v17/range_widget_horizontal.xml (hunk 4 above)
Comment (grzesiek2010) [reply collect_app/src/main/res/layout-v17/range_widget_horizontal.xml:1]: It's there https://github.com/opendatakit/collect/pull/3625/files#diff-68e08140c170fb774f8b1da8213e3fdc so seems fine.

File: collect_app/src/main/res/layout/material_full_screen_dialog_toolbar.xml [hunk 16]
```diff
@@ -13,11 +13,4 @@
         app:navigationIcon="@drawable/ic_close_black_24dp"
         tools:title="Title" />
 
-    <include
```
Comment (grzesiek2010) [inline collect_app/src/main/res/layout/material_full_screen_dialog_toolbar.xml:16]: Why did you remove it?

File: collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieActivity.java [hunk 17]
```diff
@@ -16,95 +16,36 @@
...
-            preview.setOnClickListener(new View.OnClickListener() {
-                @Override
-                public void onClick(View v) {
-                    camera.takePicture(null, null, picture);
-                }
-            });
+        requestWindowFeature(Window.FEATURE_NO_TITLE);
+        getWindow().setFlags(WindowManager
+                .LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
+        setContentView(R.layout.activity_capture_selfie_new_api);
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieActivity.java]: Please rename this layout file as well removing `_new_api`.

File: collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieVideoActivity.java [hunk 18]
```diff
@@ -17,196 +18,31 @@
...
-            try {
-                cameraId = CameraUtils.getFrontCameraId();
-                camera = CameraUtils.getCameraInstance(this, cameraId);
-            } catch (Exception e) {
-                Timber.e(e);
-            }
-
-            this.camPreview = new CameraPreview(this, camera);
-            preview.addView(this.camPreview);
+        setContentView(R.layout.activity_capture_selfie_video_new_api);
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieVideoActivity.java]: Please rename this layout file as well removing `_new_api`.

**Comment by grzesiek2010** [comment]
@SaumiaSinghal could you fix tests that failed?

**Comment by SaumiaSinghal** [comment]
Sorry @grzesiek2010  for not checking the tests earlier. Hope it fixes the test!

File: collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java [hunk 19]
```diff
@@ -223,7 +223,7 @@ private int getTheme() {
         int theme = 0;
         // https://github.com/opendatakit/collect/issues/1424
         // https://github.com/opendatakit/collect/issues/1367
-        if (!isBrokenSamsungDevice() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+        if (!isBrokenSamsungDevice()) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java:224]: There is something to improve in that method as well. Please take a look.

File: collect_app/src/main/res/layout/sort_item_layout.xml [hunk 20]
```diff
@@ -20,7 +20,6 @@ the License.
     android:layout_height="wrap_content"
     android:clickable="true"
     android:focusable="true"
-    android:foreground="?android:attr/selectableItemBackground"
```
Comment (grzesiek2010) [inline collect_app/src/main/res/layout/sort_item_layout.xml]: Why did you remove it?

File: collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java (hunk 19 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java:224]: I mean you can get rid of `&& Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP`

**Comment by grzesiek2010** [comment]
@SaumiaSinghal 
I ran into a comment in `FormsProvider` class
```
    // Leading slashes are removed from paths to support minSdkVersion < 18:
    // https://developer.android.com/reference/android/content/UriMatcher
```

I think that means we don't need `UriUtils.stripLeadingUriSlashes()` please take a look.

File: collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java (hunk 19 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java:224]: Please also get rid of those two comments with links to issues. We can find related issues using git.

**Comment by SaumiaSinghal** [comment]
Thanks @grzesiek2010 for reviewing again!

File: collect_app/src/main/res/layout/sort_item_layout.xml (hunk 20 above)
Comment (SaumiaSinghal) [reply collect_app/src/main/res/layout/sort_item_layout.xml]: Sorry @grzesiek2010! I was fixing it actually. I found it [here](https://stackoverflow.com/questions/52329249/why-doesnt-androidforeground-attribute-work) that `android:foreground` attribute doesn't work in API 23 or lower.

**Comment by SaumiaSinghal** [comment]
I also found `TODO: use just 'serverUrlPreference.getEditText().setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_arrow_drop_down, 0);' once minSdkVersion is >= 21` in `ServerPreferencesFragment.java` class. I did that too. I hope it's fine.

File: collect_app/src/main/res/layout/sort_item_layout.xml [hunk 21]
```diff
@@ -14,12 +14,13 @@ License for the specific language governing permissions and limitations under
 the License.
 
 -->
-<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (grzesiek2010) [inline collect_app/src/main/res/layout/sort_item_layout.xml]: Why `FrameLayout`?

File: collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java [hunk 22]
```diff
@@ -528,14 +528,12 @@ public int update(Uri uri, ContentValues values, String where,
         return newWhereArgs;
     }
 
-    // Leading slashes are removed from paths to support minSdkVersion < 18:
-    // https://developer.android.com/reference/android/content/UriMatcher
     static {
         URI_MATCHER = new UriMatcher(UriMatcher.NO_MATCH);
-        URI_MATCHER.addURI(FormsProviderAPI.AUTHORITY, UriUtils.stripLeadingUriSlashes(FormsColumns.CONTENT_URI.getPath()), FORMS);
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java:535]: You can now remove `UriUtils` class and the related `UriUtilsTest`.

File: collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java (hunk 19 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java:224]: Still `isBrokenSamsungDevice()` is not updated.

**Comment by grzesiek2010** [comment]
Could you also take a look at `ObjectUtils` class. Looks like we don't need it anymore.

File: collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java [hunk 23]
```diff
@@ -528,14 +528,12 @@ public int update(Uri uri, ContentValues values, String where,
...
-        URI_MATCHER.addURI(FormsProviderAPI.AUTHORITY, UriUtils.stripLeadingUriSlashes(FormsColumns.CONTENT_URI.getPath()) + "/#", FORM_ID);
+        URI_MATCHER.addURI(FormsProviderAPI.AUTHORITY, FormsColumns.CONTENT_URI.getPath(), FORMS);
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java:532]: I think it's a wrong solution please take a look at this commit where it was implemented https://github.com/opendatakit/collect/commit/f22b49d0e600824447aefd6dfa69dbf96ecddebc and InstancesProvider class.
I think it should be:
`forms`
`forms/#`
`uniqueFormsByFormId`

**Comment by grzesiek2010** [comment]
Hey @SaumiaSinghal we are really close with this pr, do you need any help to solve my last comments?

File: collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java (hunk 23 above)
Comment (SaumiaSinghal) [reply collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java:532]: Hie @grzesiek2010! Thanks for directing to the commit. I reverted it to how it was implemented.

File: collect_app/src/main/res/layout/sort_item_layout.xml (hunk 21 above)
Comment (SaumiaSinghal) [reply collect_app/src/main/res/layout/sort_item_layout.xml]: I found that `android:foreground` attribute does not work in API 23. I read about it [here](https://stackoverflow.com/questions/52329249/why-doesnt-androidforeground-attribute-work). Including it in FrameLayout works.

**Comment by SaumiaSinghal** [comment]
Hie @grzesiek2010! Sorry for fixing this late.:/

File: collect_app/src/main/res/layout/sort_item_layout.xml (hunk 21 above)
Comment (grzesiek2010) [reply collect_app/src/main/res/layout/sort_item_layout.xml]: Even if it's true it's not a related change so you can file a separate issue for that. You can't just change `LineraLayout` to `FrameLayout` because those component's work in a different way.

File: collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java (hunk 19 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java:224]: What about this @SaumiaSinghal?

**Comment by SaumiaSinghal** [comment]
Hie @grzesiek2010! Yes actually the linear layout to frame layout change is not related to this issue. So maybe, we can look into in in a separate one.

File: collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java (hunk 19 above)
Comment (SaumiaSinghal) [reply collect_app/src/main/java/org/odk/collect/android/widgets/DateWidget.java:224]: I read your comment and removed the link to those issues, but jumped `&& Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP`, I updated it.

File: collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java [hunk 24]
```diff
@@ -587,20 +584,6 @@ public static String getFileExtension(String fileName) {
     public static void grantFilePermissions(Intent intent, Uri uri, Context context) {
         intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
         intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
-
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java:590]: Please get rid of the comment above since the explanation won't be needed now.

Comment (SaumiaSinghal) [reply collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java:590]: Yes Correct!

File: collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java (hunk 23 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java:532]: I made a mistake and what you did here was right. Could you add it back?
Sorry my mistake... Please just revert this https://github.com/opendatakit/collect/pull/3625/commits/005536279258cd4dd0b51772925a9e825bf9e696

**Comment by grzesiek2010** [comment]
@seadowg I added one last comment and it should be ready. You might want to take a look since you started reviewing.

File: collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java (hunk 23 above)
Comment (SaumiaSinghal) [reply collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java:532]: Yes that's okay @grzesiek2010! I'll revert it.

File: collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieVideoActivity.java [hunk 25]
```diff
@@ -1,13 +1,14 @@
+
 package org.odk.collect.android.activities;
 
 /*
- * Copyright (C) 2013 The Android Open Source Project
+ * Copyright 2014 The Android Open Source Project
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/activities/CaptureSelfieVideoActivity.java:5]: Why do we still need this comment?

**Review by seadowg — APPROVED** [review]
One comment but looks good!

**Comment by lognaturel** [comment]
Wow, @SaumiaSinghal, this represents a lot of effort! There were more changes than I had quickly estimated from memory and I really appreciate you sticking with it and working so well with @grzesiek2010.

For quality assurance, I suggest paying close attention to Android 5.0 which is now the lowest version supported. I think media-related functionality is particularly at risk. Perhaps @grzesiek2010 could work directly with @opendatakit/testers to identify the highest-risk changes to verify before a merge. Of course, the ideal will be to merge as soon as possible so we can exercise the changes in the last couple of weeks before the release.

**Comment by SaumiaSinghal** [comment]
Thanks @lognaturel!

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java [hunk 26]
```diff
@@ -2783,11 +2782,7 @@ private void displayUIFor(@Nullable BackgroundLocationManager.BackgroundLocation
         // double space where the placeholder is in the text resource. Collapse those spaces.
         // See https://github.com/opendatakit/collect/pull/2864
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: Please remove this comment

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: I also ran into a really old comment in SounnerMultiWidget class, please remove this part in the same commit:
> There is a bug in
  android versions previous to 2.0 that affects this widget. You can find the report here:
  http://code.google.com/p/android/issues/detail?id=922 This bug causes text to be white in alert
  boxes, which makes the select options invisible in this widget. For this reason, this widget
  should not be used on phones with android versions lower than 2.0.

File: collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java [hunk 27]
```diff
@@ -275,13 +215,6 @@ public void setBinaryData(Object object) {
         // Need to have this ugly code to account for
         // a bug in the Nexus 7 on 4.3 not returning the mediaUri in the data
         // of the intent - uri in this case is a file
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java]: We also don't need this comment anymore.

Comment (SaumiaSinghal) [reply collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java]: Thanks for directing @grzesiek2010!

File: collect_app/src/main/res/layout/material_full_screen_dialog_toolbar.xml (hunk 16 above)
Comment (grzesiek2010) [reply collect_app/src/main/res/layout/material_full_screen_dialog_toolbar.xml:16]: You didn't answer my question but I think you were right you removed the entire view because now it's always gone right and not needed? If that's true please remove it. 

It was just a question because as a reviewer I didn't catch it at first glance, don't hesitate to explain or even argue if you are sure you are right :smiley: just a tip for the future.

**Comment by grzesiek2010** [comment]
The last thing from me @SaumiaSinghal https://github.com/opendatakit/collect/pull/3625/files#r383858651

File: collect_app/src/main/res/layout/material_full_screen_dialog_toolbar.xml (hunk 16 above)
Comment (SaumiaSinghal) [reply collect_app/src/main/res/layout/material_full_screen_dialog_toolbar.xml:16]: Hi @grzesiek2010! I removed it because I noticed in the tests that it was not needed now. Thank you for the tip! :D 
I'll fix it.

**Review by grzesiek2010 — APPROVED** [review]

**Comment by mmarciniak90** [comment]
**Tested with success**

I worked with @kkrawczyk123 to find some regression.
Verified on Android: 5.1, 6.0, 7.0, 8.1, 9.0, 10.0

Android 4.x doesn't allow installing this version. 
Error is visible:


Verified cases:
- media from form
- media attached to the submission
- versioning
- audit
- Aggregate, Central, Google Drive
- external CSV
- autoplay
- etc.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Comment by grzesiek2010** [comment]
Thanks @SaumiaSinghal and congrats you made the code cleaner.
